src/Evolution/Systems/Cce/Components/CharacteristicEvolution.hpp

```diff
@@ -89,15 +89,8 @@ struct CharacteristicEvolution {
   using chare_type = Parallel::Algorithms::Singleton;
   using metavariables = Metavariables;
 
-  using initialize_action_list =
-      tmpl::list<Actions::InitializeCharacteristicEvolutionVariables,
-                 Actions::InitializeCharacteristicEvolutionTime,
-                 Actions::InitializeCharacteristicEvolutionScri,
-                 Actions::RequestBoundaryData<
-                     typename Metavariables::cce_boundary_component,
-                     CharacteristicEvolution<Metavariables>>,
-                 Actions::ReceiveWorldtubeData<Metavariables>,
-                 ::Actions::MutateApply<
+  using mutate_apply_initialize_first_hypersurface =
```

Comment:
no, we really want one action that does what all the chained MutateApply's do.

Reply:
Would you mind explaining why it is so strongly preferred? It has the cursory appearance of preferring code and test duplication.

Reply:
We prefer one action to do multiple things instead of chaining multiple actions that mutate one thing in the DataBox.    In general, if no communication is being done between actions, then those actions should just be combined into one action calling multiple functions.    One option could be to have a MutateApply that mutates multiple things, but one concern to watch out for is whether doing so causes you to put more things in the DataBox than you really need to put in.  In other words, if the chain you have has some initial inputs and computes some outputs, are there temporary quantities in the DataBox that only exist because you are mutating one thing at a time.

Reply:
I see, the suggested design is intended to avoid the buildup of temporaries in the databox? I suppose I could omit the `CauchyAngularCoords` from the databox in each of these chains in favor of allocating the tensor in both of the new actions and making the function calls more ugly by needing to specify all of the `apply` arguments explicitly.

So is it against the philosophy of this request to create an action `MutateApplyAll` taking a `tmpl::list` of invokables and just calls `mutate_apply` with each of the invokables in sequence on the databox?

Reply:
In general a `MutateApplyAll` is not ideal, but I'm fine doing that in the interest of getting this merged sooner rather than later :)

Reply:
Okay; do I understand correctly then that the real preference is toward larger actions fully specialized to their task; and if that task happens to be several `mutate_apply`s, then it should do that within its `apply`, but a long action list is undesirable?
I'm just genuinely trying to understand the current ideas for best-practices for action lists, because the request to split up the Initialization action in the previous PR seems at odds with a best-practice idea of keeping actions as bundled as can be practically achieved.

Reply:
Ah, okay, I see. Yes, for initialization actions we've found that a lot of them can be reused across different evolution systems or even shared with the elliptic solver. Additionally, since we only do initialization once, there's a lot less need for optimizations that are only possible when you know exactly what system you're solving, etc. Since the initialization actions also add things for the evolution, if temporaries are removed from the evolution, the associated initialization code also disappears.

For the evolution part I (we) think it's better to have fewer actions that do a lot of work, are optimized to reduce allocations while maintaining a small memory footprint, and are even specialized for systems. An example is for DG-subcell it'll be a lot easier to have a large RHS, slice, send, compute volume terms, etc. action than many small ones because the code is specialized to DG-subcell, and so won't generically be portable anyway, and we can write free functions if it happens to be portable.

I've generally also found that as code becomes more spread out and is only tied together in a super generic "Tie things" structure it becomes harder to follow control flow or to search for things in the code. This is a big reason why I think we should move to making some of these wrappers around wrappers go away.

Does that help clarify where we're coming from?

Reply:
Yeah, that helps. It's certainly true that most of the CCE hypersurface procedure is similarly non-generalizable.

Reply:
Yes, but I think it's better to deal with that sort of possible refactors after this is merged and some performance measurements can be made :)